pkg/adapter/mtping/runner.go

```diff
@@ -140,16 +132,44 @@ func (a *cronJobsRunner) cronTick(ctx context.Context, event cloudevents.Event)
 	}
 }
 
-type message struct {
-	Body string `json:"body"`
-}
+func makeEvent(source *sourcesv1beta2.PingSource) cloudevents.Event {
+	event := cloudevents.NewEvent()
+	event.SetType(sourcesv1beta2.PingSourceEventType)
+	event.SetSource(sourcesv1beta2.PingSourceSource(source.Namespace, source.Name))
+	if source.Spec.CloudEventOverrides != nil && source.Spec.CloudEventOverrides.Extensions != nil {
+		for key, override := range source.Spec.CloudEventOverrides.Extensions {
+			event.SetExtension(key, override)
+		}
+	}
 
-func makeMessage(body string) interface{} {
-	// try to marshal the body into an interface.
-	var objmap map[string]*json.RawMessage
-	if err := json.Unmarshal([]byte(body), &objmap); err != nil {
-		// Default to a wrapped message.
-		return message{Body: body}
+	// Set event data, at most one of data and dataBase64 exists.
+	// 1. If dataBase64 exists, then it's binary data, set event.DataEncoded to dataBase64.
+	// 2. If data exists, then it's not binary data
+	//    a. If contentType is not `application/json`, set event.DataEncoded to data
+	//    b. If contentType is `application/json`:
+	//   	 - If data is valid JSON, unmarshal it into an interface, event.DataEncoded will be json.Marshal(data)
+	//       - Otherwise, make sure event.DataEncoded is the same as data.
+	if source.Spec.DataBase64 != "" {
+		event.SetData(source.Spec.ContentType, []byte(source.Spec.DataBase64))
+	} else {
+		switch source.Spec.ContentType {
+		// TODO: Do we need to deal with `application/cloudevents+json` and `application/cloudevents-batch+json`?
+		case cloudevents.ApplicationJSON:
+			// try to unmarshal the body into an interface.
+			var objmap map[string]*json.RawMessage
+			err := json.Unmarshal([]byte(source.Spec.Data), &objmap)
+
+			if err != nil {
+				// cannot be unmarshalled, we pass it as is and bypass the json encoding
+				event.SetData(cloudevents.TextPlain, source.Spec.Data)
+				event.SetDataContentType(source.Spec.ContentType)
```

Comment:
Is this necessary since the webhook rejects such configuration?

Reply:
Thanks, removed this part.